## Data Wrangle WeRateDogs Twitter Feed
By Ryan Gee

In [1]:
# Import Required Packages
import os
import sys
import numpy as np
import pandas as pd
from sqlalchemy import create_engine
module_path = os.path.abspath(os.path.join('..'))
if module_path not in sys.path:
    sys.path.append(module_path)

In [2]:
# Custom Packages
from models.we_rate_dogs_data_puller import WeRateDogsDataPuller

### Gather Data

##### 1. Get Archived We Rate Dogs Twitter Data
Data downloaded from [Udacity Course.](https://d17h27t6h515a5.cloudfront.net/topher/2017/August/59a4e958_twitter-archive-enhanced/twitter-archive-enhanced.csv)

In [3]:
dogs = WeRateDogsDataPuller()

In [4]:
dogs.get_archive_data().head(3)

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
0  892420643555336193                    NaN                  NaN   
1  892177421306343426                    NaN                  NaN   
2  891815181378084864                    NaN                  NaN   

                   timestamp  \
0  2017-08-01 16:23:56 +0000   
1  2017-08-01 00:17:27 +0000   
2  2017-07-31 00:18:03 +0000   

                                              source  \
0  <a href="http://twitter.com/download/iphone" r...   
1  <a href="http://twitter.com/download/iphone" r...   
2  <a href="http://twitter.com/download/iphone" r...   

                                                text  retweeted_status_id  \
0  This is Phineas. He's a mystical boy. Only eve...                  NaN   
1  This is Tilly. She's just checking pup on you....                  NaN   
2  This is Archie. He is a rare Norwegian Pouncin...                  NaN   

   retweeted_status_user_id retweeted_status_timestamp  \
0                       NaN                        NaN   
1                       NaN                        NaN   
2                       NaN                        NaN   

                                       expanded_urls  rating_numerator  \
0  https://twitter.com/dog_rates/status/892420643...                13   
1  https://twitter.com/dog_rates/status/892177421...                13   
2  https://twitter.com/dog_rates/status/891815181...                12   

   rating_denominator     name doggo floofer pupper puppo  
0                  10  Phineas  None    None   None  None  
1                  10    Tilly  None    None   None  None  
2                  10   Archie  None    None   None  None

##### 2. Get We Rate Dogs Image Prediction Data
Use the Requests library to programatically download data from a tsv file on hosted server.

In [5]:
dogs.get_image_data().head(3)

tweet_id                                          jpg_url  \
0  666020888022790149  https://pbs.twimg.com/media/CT4udn0WwAA0aMy.jpg   
1  666029285002620928  https://pbs.twimg.com/media/CT42GRgUYAA5iDo.jpg   
2  666033412701032449  https://pbs.twimg.com/media/CT4521TWwAEvMyu.jpg   

   img_num                      p1   p1_conf  p1_dog                  p2  \
0        1  Welsh_springer_spaniel  0.465074    True              collie   
1        1                 redbone  0.506826    True  miniature_pinscher   
2        1         German_shepherd  0.596461    True            malinois   

    p2_conf  p2_dog                   p3   p3_conf  p3_dog  
0  0.156665    True    Shetland_sheepdog  0.061428    True  
1  0.074192    True  Rhodesian_ridgeback  0.072010    True  
2  0.138584    True           bloodhound  0.116197    True

##### 3. Get Retweet and Favorite count data
Use Twitter's API to gather retweet and favorite counts for each dog and write it to a csv file.

In [6]:
# dogs.write_twitter_data_to_file()

Read the new csv file and convert to a Panda's Dataframe.

In [7]:
dogs.get_twitter_data_as_df().head(3)

id  retweet_count  favorite_count
0  892420643555336193           8346           38152
1  892177421306343426           6169           32719
2  891815181378084864           4082           24637

##### 4. Merge all three datasets into one dataset

In [8]:
dogs.get_combined_data().head(5)

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
0  892420643555336193                    NaN                  NaN   
1  892177421306343426                    NaN                  NaN   
2  891815181378084864                    NaN                  NaN   
3  891689557279858688                    NaN                  NaN   
4  891327558926688256                    NaN                  NaN   

                   timestamp  \
0  2017-08-01 16:23:56 +0000   
1  2017-08-01 00:17:27 +0000   
2  2017-07-31 00:18:03 +0000   
3  2017-07-30 15:58:51 +0000   
4  2017-07-29 16:00:24 +0000   

                                              source  \
0  <a href="http://twitter.com/download/iphone" r...   
1  <a href="http://twitter.com/download/iphone" r...   
2  <a href="http://twitter.com/download/iphone" r...   
3  <a href="http://twitter.com/download/iphone" r...   
4  <a href="http://twitter.com/download/iphone" r...   

                                                text  retweeted_status_id  \
0  This is Phineas. He's a mystical boy. Only eve...                  NaN   
1  This is Tilly. She's just checking pup on you....                  NaN   
2  This is Archie. He is a rare Norwegian Pouncin...                  NaN   
3  This is Darla. She commenced a snooze mid meal...                  NaN   
4  This is Franklin. He would like you to stop ca...                  NaN   

   retweeted_status_user_id retweeted_status_timestamp  \
0                       NaN                        NaN   
1                       NaN                        NaN   
2                       NaN                        NaN   
3                       NaN                        NaN   
4                       NaN                        NaN   

                                       expanded_urls   ...    img_num  \
0  https://twitter.com/dog_rates/status/892420643...   ...        1.0   
1  https://twitter.com/dog_rates/status/892177421...   ...        1.0   
2  https://twitter.com/dog_rates/status/891815181...   ...        1.0   
3  https://twitter.com/dog_rates/status/891689557...   ...        1.0   
4  https://twitter.com/dog_rates/status/891327558...   ...        2.0   

            p1   p1_conf p1_dog                  p2   p2_conf p2_dog  \
0       orange  0.097049  False               bagel  0.085851  False   
1    Chihuahua  0.323581   True            Pekinese  0.090647   True   
2    Chihuahua  0.716012   True            malamute  0.078253   True   
3  paper_towel  0.170278  False  Labrador_retriever  0.168086   True   
4       basset  0.555712   True    English_springer  0.225770   True   

                            p3   p3_conf  p3_dog  
0                       banana  0.076110   False  
1                     papillon  0.068957    True  
2                       kelpie  0.031379    True  
3                      spatula  0.040836   False  
4  German_short-haired_pointer  0.175219    True  

[5 rows x 31 columns]

### Assess Dataset for Errors
Assess the data visually in Excel and then programatically using Panda's functions. The quality and tidiness issues are listed at the bottom of this section.

In [9]:
df = dogs.get_combined_data()
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2356 entries, 0 to 2355
Data columns (total 31 columns):
tweet_id                      2356 non-null int64
in_reply_to_status_id         78 non-null float64
in_reply_to_user_id           78 non-null float64
timestamp                     2356 non-null object
source                        2356 non-null object
text                          2356 non-null object
retweeted_status_id           181 non-null float64
retweeted_status_user_id      181 non-null float64
retweeted_status_timestamp    181 non-null object
expanded_urls                 2297 non-null object
rating_numerator              2356 non-null int64
rating_denominator            2356 non-null int64
name                          2356 non-null object
doggo                         2356 non-null object
floofer                       2356 non-null object
pupper                        2356 non-null object
puppo                         2356 non-null object
id                            23

In [10]:
df.describe()

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
count  2.356000e+03           7.800000e+01         7.800000e+01   
mean   7.427716e+17           7.455079e+17         2.014171e+16   
std    6.856705e+16           7.582492e+16         1.252797e+17   
min    6.660209e+17           6.658147e+17         1.185634e+07   
25%    6.783989e+17           6.757419e+17         3.086374e+08   
50%    7.196279e+17           7.038708e+17         4.196984e+09   
75%    7.993373e+17           8.257804e+17         4.196984e+09   
max    8.924206e+17           8.862664e+17         8.405479e+17   

       retweeted_status_id  retweeted_status_user_id  rating_numerator  \
count         1.810000e+02              1.810000e+02       2356.000000   
mean          7.720400e+17              1.241698e+16         13.126486   
std           6.236928e+16              9.599254e+16         45.876648   
min           6.661041e+17              7.832140e+05          0.000000   
25%           7.186315e+17              4.196984e+09         10.000000   
50%           7.804657e+17              4.196984e+09         11.000000   
75%           8.203146e+17              4.196984e+09         12.000000   
max           8.874740e+17              7.874618e+17       1776.000000   

       rating_denominator            id  retweet_count  favorite_count  \
count         2356.000000  2.342000e+03    2342.000000     2342.000000   
mean            10.455433  7.422212e+17    2943.696413     7985.288642   
std              6.745237  6.832408e+16    4950.394267    12359.959313   
min              0.000000  6.660209e+17       0.000000        0.000000   
25%             10.000000  6.783509e+17     591.500000     1377.500000   
50%             10.000000  7.186224e+17    1374.500000     3473.500000   
75%             10.000000  7.986971e+17    3432.250000     9782.250000   
max            170.000000  8.924206e+17   83905.000000   164672.000000   

           img_num      p1_conf       p2_conf       p3_conf  
count  2075.000000  2075.000000  2.075000e+03  2.075000e+03  
mean      1.203855     0.594548  1.345886e-01  6.032417e-02  
std       0.561875     0.271174  1.006657e-01  5.090593e-02  
min       1.000000     0.044333  1.011300e-08  1.740170e-10  
25%       1.000000     0.364412  5.388625e-02  1.622240e-02  
50%       1.000000     0.588230  1.181810e-01  4.944380e-02  
75%       1.000000     0.843855  1.955655e-01  9.180755e-02  
max       4.000000     1.000000  4.880140e-01  2.734190e-01

The max rating denominator is 170, so let's see how many denominators are greater than 10. These high denominators may be due to errors in the the parsing logic that extracts the ratings.

In [11]:
df[df.rating_denominator > 10][['text','rating_numerator','rating_denominator']]

text  rating_numerator  \
342            @docmisterio account started on 11/15/15                11   
433   The floofs have been released I repeat the flo...                84   
784   RT @dog_rates: After so many requests, this is...                 9   
902   Why does this never happen at my front door......               165   
1068  After so many requests, this is Bretagne. She ...                 9   
1120  Say hello to this unbelievably well behaved sq...               204   
1165  Happy 4/20 from the squad! 13/10 for all https...                 4   
1202  This is Bluebert. He just saw that both #Final...                50   
1228  Happy Saturday here's 9 puppers on a bench. 99...                99   
1254  Here's a brigade of puppers. All look very pre...                80   
1274  From left to right:\nCletus, Jerome, Alejandro...                45   
1351  Here is a whole flock of puppers.  60/50 I'll ...                60   
1433  Happy Wednesday here's a bucket of pups. 44/40...                44   
1598  Yes I do realize a rating of 4/20 would've bee...                 4   
1634  Two sneaky puppers were not initially seen, mo...               143   
1635  Someone help the girl is being mugged. Several...               121   
1662  This is Darrel. He just robbed a 7/11 and is i...                 7   
1663  I'm aware that I could've said 20/16, but here...                20   
1779  IT'S PUPPERGEDDON. Total of 144/120 ...I think...               144   
1843  Here we have an entire platoon of puppers. Tot...                88   

      rating_denominator  
342                   15  
433                   70  
784                   11  
902                  150  
1068                  11  
1120                 170  
1165                  20  
1202                  50  
1228                  90  
1254                  80  
1274                  50  
1351                  50  
1433                  40  
1598                  20  
1634                 130  
1635                 110  
1662                  11  
1663                  16  
1779                 120  
1843                  80

In [12]:
# Check format of source
df.source.value_counts()

<a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>     2221
<a href="http://vine.co" rel="nofollow">Vine - Make a Scene</a>                          91
<a href="http://twitter.com" rel="nofollow">Twitter Web Client</a>                       33
<a href="https://about.twitter.com/products/tweetdeck" rel="nofollow">TweetDeck</a>      11
Name: source, dtype: int64

In [13]:
df.name.value_counts()[:15]

None       745
a           55
Charlie     12
Lucy        11
Cooper      11
Oliver      11
Penny       10
Lola        10
Tucker      10
Winston      9
Bo           9
the          8
Sadie        8
an           7
Daisy        7
Name: name, dtype: int64

In [14]:
# Find all invalid Dog Names
df[df.name.apply(lambda x: x[0].isupper()==False)].name.value_counts()

a               55
the              8
an               7
very             5
just             4
quite            4
one              4
not              2
getting          2
mad              2
actually         2
light            1
my               1
this             1
life             1
infuriating      1
his              1
officially       1
incredibly       1
unacceptable     1
old              1
such             1
space            1
by               1
all              1
Name: name, dtype: int64

#### Visual Assesment
**`Quality Issues`**:
    1. _Some expanded_urls have multiple gofundme and twitter links that are duplicates
    2. _Source needs to be parsed into categories like iphone, web client, etc.
    3. _Some names are missing or invalid dog names like 'the', 'a', and 'an'.
    4. _Dog breeds in the image prediction data are not consistant. Some are capitalized and some are not.
    5. _Missing data for retweet counts, favorite counts, in_reply_to_status_id, and in_reply_to_user_id.
    6. _Numerators with a decimal are parsed incorrectly. (ex: 9.75/10 is parsed as 75/10)
**`Tidiness Issues`**:
    1. _Doggo, floofer, pupper, and puppo columns need to be melted into one colum called dog_type.
    2. _Duplicated columns: tweet_id and id.
    3. _The image predictions table need to be seperated into a different table.
    4. _The columns in the image prediction table need to be melted into 5 columns: tweet_id, prediction_rank, prediction, prediction_confidence, and is_dog.

#### Programatic Assesment
** `Quality Issues` **
    1. _timestamp and retweeted_status_timestamp columns are strings, but should be datetimes.
    2. _retweet_count, favorite_count, retweeted_status_id and retweeted_status_user_id should be int64 not float64.
    3. _There are rating_denominators greater than 10.

### Clean Dataset of Errors


**`Tidiness Issues`**

* Duplicated columns: tweet_id and id.
* The image predictions table need to be seperated into a different table.

**Define**

Separate the first 17 fields into a dataframe called 'tweets' and the last fields go into a dataframe called 'predictions'.

**Code**

In [15]:
#create a copy of the dataframe to clean and tidy
df_clean = df.copy()

In [16]:
#Separate into two tables
tweets_clean = df_clean.loc[:, :'favorite_count'].drop(['id'], axis=1)
predictions_clean = dogs.get_image_data()

**Test**

In [17]:
tweets_clean.head(2)

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
0  892420643555336193                    NaN                  NaN   
1  892177421306343426                    NaN                  NaN   

                   timestamp  \
0  2017-08-01 16:23:56 +0000   
1  2017-08-01 00:17:27 +0000   

                                              source  \
0  <a href="http://twitter.com/download/iphone" r...   
1  <a href="http://twitter.com/download/iphone" r...   

                                                text  retweeted_status_id  \
0  This is Phineas. He's a mystical boy. Only eve...                  NaN   
1  This is Tilly. She's just checking pup on you....                  NaN   

   retweeted_status_user_id retweeted_status_timestamp  \
0                       NaN                        NaN   
1                       NaN                        NaN   

                                       expanded_urls  rating_numerator  \
0  https://twitter.com/dog_rates/status/892420643...                13   
1  https://twitter.com/dog_rates/status/892177421...                13   

   rating_denominator     name doggo floofer pupper puppo  retweet_count  \
0                  10  Phineas  None    None   None  None         8346.0   
1                  10    Tilly  None    None   None  None         6169.0   

   favorite_count  
0         38152.0  
1         32719.0

In [18]:
predictions_clean.head(2)

tweet_id                                          jpg_url  \
0  666020888022790149  https://pbs.twimg.com/media/CT4udn0WwAA0aMy.jpg   
1  666029285002620928  https://pbs.twimg.com/media/CT42GRgUYAA5iDo.jpg   

   img_num                      p1   p1_conf  p1_dog                  p2  \
0        1  Welsh_springer_spaniel  0.465074    True              collie   
1        1                 redbone  0.506826    True  miniature_pinscher   

    p2_conf  p2_dog                   p3   p3_conf  p3_dog  
0  0.156665    True    Shetland_sheepdog  0.061428    True  
1  0.074192    True  Rhodesian_ridgeback  0.072010    True

**`Tidiness Issues in Predictions Dataframe`**

* The columns in the 'predictions' dataframe need to be melted into 5 columns: tweet_id, prediction_order, prediction, prediction_confidence, and is_dog.

**Define**

Melt p1, p2, and p3 columns in the predictions table into a column called prediction_rank (with values 1, 2, or 3) and a column called prediction (with the dog breed prediction). Then melt p1_conf, p2_conf, and p3_conf into a column called prediction_confidence. Finally melt p1_dog, p2_dog, and p3_dog into a column called is_dog.

**Code**

In [19]:
prediction = pd.melt(predictions_clean, id_vars=['tweet_id','jpg_url','img_num'], value_vars=['p1','p2','p3'],
              var_name='prediction_rank', value_name='prediction')

In [20]:
prediction_conf = pd.melt(predictions_clean, id_vars=['tweet_id'], value_vars=['p1_conf','p2_conf','p3_conf'],
              var_name='prediction_rank', value_name='prediction_confidence').prediction_confidence

In [21]:
is_dog = pd.melt(predictions_clean, id_vars=['tweet_id'], value_vars=['p1_dog','p2_dog','p3_dog'],
              var_name='prediction_rank', value_name='is_dog').is_dog

In [22]:
predictions_clean = (prediction.join(prediction_conf)).join(is_dog)

**Test**

In [23]:
predictions_clean[predictions_clean.tweet_id.isin([666020888022790149,666029285002620928])].sort_values('tweet_id')

tweet_id                                          jpg_url  \
0     666020888022790149  https://pbs.twimg.com/media/CT4udn0WwAA0aMy.jpg   
2075  666020888022790149  https://pbs.twimg.com/media/CT4udn0WwAA0aMy.jpg   
4150  666020888022790149  https://pbs.twimg.com/media/CT4udn0WwAA0aMy.jpg   
1     666029285002620928  https://pbs.twimg.com/media/CT42GRgUYAA5iDo.jpg   
2076  666029285002620928  https://pbs.twimg.com/media/CT42GRgUYAA5iDo.jpg   
4151  666029285002620928  https://pbs.twimg.com/media/CT42GRgUYAA5iDo.jpg   

      img_num prediction_rank              prediction  prediction_confidence  \
0           1              p1  Welsh_springer_spaniel               0.465074   
2075        1              p2                  collie               0.156665   
4150        1              p3       Shetland_sheepdog               0.061428   
1           1              p1                 redbone               0.506826   
2076        1              p2      miniature_pinscher               0.074192   
4151        1              p3     Rhodesian_ridgeback               0.072010   

      is_dog  
0       True  
2075    True  
4150    True  
1       True  
2076    True  
4151    True

In [24]:
# Check if there are any NaN tweet_id's
sum(predictions_clean.tweet_id.isna())

0

In [25]:
predictions_clean.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6225 entries, 0 to 6224
Data columns (total 7 columns):
tweet_id                 6225 non-null int64
jpg_url                  6225 non-null object
img_num                  6225 non-null int64
prediction_rank          6225 non-null object
prediction               6225 non-null object
prediction_confidence    6225 non-null float64
is_dog                   6225 non-null bool
dtypes: bool(1), float64(1), int64(2), object(3)
memory usage: 298.0+ KB


Okay, the Predictions Dataframe looks tidy. Let's move on to cleaning this Dataframe.

**`Quality Issues in Predictions Dataframe`**

* Dog breeds in the image prediction data are not consistant. Some are capitalized and some are not.

**Define**

Standardize capitalization for dog breeds using .lower() function.

**Code**

In [26]:
# Create copy of pre cleaned dataframe for testing purposes
pre_clean_copy = predictions_clean.copy()
# Clean the dataframe
predictions_clean['prediction'] = predictions_clean['prediction'].apply(lambda x: x.lower())

**Test**

In [27]:
# Find the duplicate predictions (aka same predictions with different capitalization) in the pre-cleaned dataframe
l =list(map(lambda x: x.lower(), list(pre_clean_copy.prediction.value_counts().index)))
set([x for x in l if l.count(x) > 1])

{'cardigan'}

In [28]:
# Verify the duplicates in the pre-cleaned dataframe
pre_clean_copy[pre_clean_copy.prediction.isin(['cardigan', 'Cardigan'])].prediction.value_counts()

Cardigan    115
cardigan      1
Name: prediction, dtype: int64

In [29]:
# Verify that the cleaned dataframe has all lower case predicitions,
predictions_clean.prediction.value_counts().head(5)

golden_retriever      290
labrador_retriever    283
chihuahua             185
pembroke              143
cardigan              116
Name: prediction, dtype: int64

In [30]:
# Verify the duplicates have also been removed.
predictions_clean[predictions_clean.prediction.isin(['cardigan', 'Cardigan'])].prediction.value_counts()

cardigan    116
Name: prediction, dtype: int64

**`Quality Issues in Tweet Data`**

* Missing retweet counts, favorite counts, in_reply_to_status_id, in_reply_to_user_id, retweeted_status_id, retweeted_status_user_id, and retweeted_status_timestamp

**Define**

To focus on just original tweets, we will remove all rows that have data for in_reply_to_status_id, in_reply_to_user_id, retweeted_status_id, retweeted_status_user_id, or retweeted_status_timestamp. Once those rows are removed, we will drop those columns since they will all be NaN.

Finally, remove rows that don't have retweet counts or favorite counts.

**Code**

In [31]:
# create copy for testing
pre_clean_copy = tweets_clean.copy()
# Remove retweet and reply rows
tweets_clean = tweets_clean[tweets_clean.in_reply_to_status_id.isna()]
tweets_clean = tweets_clean[tweets_clean.retweeted_status_id.isna()]

In [32]:
# Remove rows with missing favorite and retweet counts
missing_count_rows = tweets_clean[tweets_clean.retweet_count.isna()].shape[0]
tweets_clean = tweets_clean[~tweets_clean.retweet_count.isna()]
tweets_clean = tweets_clean[~tweets_clean.favorite_count.isna()]

**Test**

In [33]:
print("Pre-Cleaned Copy has {} in_reply rows.".format(pre_clean_copy[~pre_clean_copy.in_reply_to_status_id.isna()].shape[0]))
print("Pre-Cleaned Copy has {} retweeted rows.".format(pre_clean_copy[~pre_clean_copy.retweeted_status_id.isna()].shape[0]))
print("Cleaned Copy has {} in_reply rows.".format(tweets_clean[~tweets_clean.in_reply_to_status_id.isna()].shape[0]))
print("Cleaned Copy has {} retweeted rows.".format(tweets_clean[~tweets_clean.retweeted_status_id.isna()].shape[0]))
print("Rows with missing Favorite and Retweet counts: {}.".format(missing_count_rows))
print("------------------------")
print("Pre-Cleaned Copy has {} total rows.".format(pre_clean_copy.shape[0]))
print("Cleaned Copy has {} total rows.".format(tweets_clean.shape[0]))

Pre-Cleaned Copy has 78 in_reply rows.
Pre-Cleaned Copy has 181 retweeted rows.
Cleaned Copy has 0 in_reply rows.
Cleaned Copy has 0 retweeted rows.
Rows with missing Favorite and Retweet counts: 1.
------------------------
Pre-Cleaned Copy has 2356 total rows.
Cleaned Copy has 2096 total rows.


**Code**

In [34]:
# Remove Columns
tweets_clean.drop(columns=['in_reply_to_status_id', 'in_reply_to_user_id', 'retweeted_status_id', 'retweeted_status_user_id', 'retweeted_status_timestamp'], inplace=True)

**Test**

In [35]:
list(tweets_clean)

['tweet_id',
 'timestamp',
 'source',
 'text',
 'expanded_urls',
 'rating_numerator',
 'rating_denominator',
 'name',
 'doggo',
 'floofer',
 'pupper',
 'puppo',
 'retweet_count',
 'favorite_count']

In [36]:
tweets_clean.head()

tweet_id                  timestamp  \
0  892420643555336193  2017-08-01 16:23:56 +0000   
1  892177421306343426  2017-08-01 00:17:27 +0000   
2  891815181378084864  2017-07-31 00:18:03 +0000   
3  891689557279858688  2017-07-30 15:58:51 +0000   
4  891327558926688256  2017-07-29 16:00:24 +0000   

                                              source  \
0  <a href="http://twitter.com/download/iphone" r...   
1  <a href="http://twitter.com/download/iphone" r...   
2  <a href="http://twitter.com/download/iphone" r...   
3  <a href="http://twitter.com/download/iphone" r...   
4  <a href="http://twitter.com/download/iphone" r...   

                                                text  \
0  This is Phineas. He's a mystical boy. Only eve...   
1  This is Tilly. She's just checking pup on you....   
2  This is Archie. He is a rare Norwegian Pouncin...   
3  This is Darla. She commenced a snooze mid meal...   
4  This is Franklin. He would like you to stop ca...   

                                       expanded_urls  rating_numerator  \
0  https://twitter.com/dog_rates/status/892420643...                13   
1  https://twitter.com/dog_rates/status/892177421...                13   
2  https://twitter.com/dog_rates/status/891815181...                12   
3  https://twitter.com/dog_rates/status/891689557...                13   
4  https://twitter.com/dog_rates/status/891327558...                12   

   rating_denominator      name doggo floofer pupper puppo  retweet_count  \
0                  10   Phineas  None    None   None  None         8346.0   
1                  10     Tilly  None    None   None  None         6169.0   
2                  10    Archie  None    None   None  None         4082.0   
3                  10     Darla  None    None   None  None         8480.0   
4                  10  Franklin  None    None   None  None         9182.0   

   favorite_count  
0         38152.0  
1         32719.0  
2         24637.0  
3         41494.0  
4         39658.0

**`Quality Issues in Tweet Data`**

* Source needs to be parsed into categories like iphone, web client, etc.

**Define**

Use regex or a string function to extract the correct source type.

**Code**

In [37]:
def parse_source(src):
    sources = ['iPhone', 'Vine', 'Web Client', 'TweetDeck']
    for s in sources:
        if s in src:
            return s

tweets_clean.source = tweets_clean.source.apply(parse_source)

**Test**

In [38]:
tweets_clean.source.value_counts()

iPhone        1963
Vine            91
Web Client      31
TweetDeck       11
Name: source, dtype: int64

**`Quality Issues in Tweet Data`**

* Some names are 'None' or invalid dog names like 'the', 'a', 'an', etc.

**Define**

Change all invalid names to 'Unknown'.

**Code**

In [39]:
# Change invalid names to None
tweets_clean.name = tweets_clean.name.apply(lambda x: x if x[0].isupper() else 'None')
# Change all Nones to Unknown
tweets_clean.name = tweets_clean.name.apply(lambda x: 'Unknown' if x == 'None' else x)

**Test**

In [40]:
tweets_clean.name.value_counts().head(5)

Unknown    707
Lucy        11
Charlie     10
Cooper      10
Oliver      10
Name: name, dtype: int64

**`Quality Issues in Tweet Data`**

* Some expanded_urls have multiple gofundme and twitter links that are duplicates

**Define**

Use the split function to break up the links into a list. Then change the list to a set in order to remove all duplicates. If there is a gofundme page put that link in a new column called gofundme_url. If the linke is from Twitter then put that link in a column called twitter_url. Finally, drop the expanded_urls column.

**Code**

In [41]:
def extract_urls(urls, keyword):
    try:
        return [link for link in set(urls.split(',')) if keyword in link][0]
    except:
        return None

tweets_clean['gofundme_url'] = tweets_clean.expanded_urls.apply(lambda x: extract_urls(str(x),'gofundme'))
tweets_clean['twitter_url'] = tweets_clean.expanded_urls.apply(lambda x: extract_urls(str(x),'twitter'))
tweets_clean.drop(columns=['expanded_urls'], inplace=True)

**Test**

In [42]:
# Check the new gofundme and twitter_url columns are added
list(tweets_clean)

['tweet_id',
 'timestamp',
 'source',
 'text',
 'rating_numerator',
 'rating_denominator',
 'name',
 'doggo',
 'floofer',
 'pupper',
 'puppo',
 'retweet_count',
 'favorite_count',
 'gofundme_url',
 'twitter_url']

In [43]:
# Check that the gofundme and twitter links are in separate columns and that the expanded_urls column is gone
tweets_clean[tweets_clean.tweet_id == 890971913173991426]

tweet_id                  timestamp  source  \
6  890971913173991426  2017-07-28 16:27:12 +0000  iPhone   

                                                text  rating_numerator  \
6  Meet Jax. He enjoys ice cream so much he gets ...                13   

   rating_denominator name doggo floofer pupper puppo  retweet_count  \
6                  10  Jax  None    None   None  None         2028.0   

   favorite_count                                 gofundme_url  \
6         11642.0  https://gofundme.com/ydvmve-surgery-for-jax   

                                         twitter_url  
6  https://twitter.com/dog_rates/status/890971913...

**`Tidy Issues in Tweet Data`**

* Doggo, floofer, pupper, and puppo columns need to be melted into one colum called dog_type.

**Define**

Use the stack and groupby function to get the non-NaN values in the doggo, floofer, pupper, and puppo columns. Create a new column called dog_type that contains these values. Finally, drop the 4 columns and leave just the dog_type column.

**Code**

In [44]:
dog_type = tweets_clean[['doggo', 'floofer', 'pupper', 'puppo']].replace('None',np.nan).stack().groupby(level=0).agg(','.join)
tweets_clean = tweets_clean.join(pd.DataFrame(dog_type,columns=['dog_type']), how='left')

In [45]:
tweets_clean.drop(columns=['doggo', 'floofer', 'pupper', 'puppo'], inplace=True)

**Test**

In [46]:
tweets_clean.dog_type.value_counts()

pupper           221
doggo             72
puppo             23
floofer            9
doggo,pupper       9
doggo,puppo        1
doggo,floofer      1
Name: dog_type, dtype: int64

**`Quality Issues in Tweet Data`**

* timestamp column is string, but should be datetime.
* retweet_count and favorite_count should be int64 not float64.

**Define**

Use the to_datetime function to convert the timestamp strings to datetime.
Use the astype function to convert the count columns to int64.

**Code**

In [47]:
tweets_clean.timestamp = pd.to_datetime(tweets_clean['timestamp'])
tweets_clean.retweet_count = tweets_clean['retweet_count'].astype(np.int64)
tweets_clean.favorite_count = tweets_clean['favorite_count'].astype(np.int64)

**Test**

In [48]:
tweets_clean.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2096 entries, 0 to 2355
Data columns (total 12 columns):
tweet_id              2096 non-null int64
timestamp             2096 non-null datetime64[ns]
source                2096 non-null object
text                  2096 non-null object
rating_numerator      2096 non-null int64
rating_denominator    2096 non-null int64
name                  2096 non-null object
retweet_count         2096 non-null int64
favorite_count        2096 non-null int64
gofundme_url          22 non-null object
twitter_url           1999 non-null object
dog_type              336 non-null object
dtypes: datetime64[ns](1), int64(5), object(6)
memory usage: 292.9+ KB


**`Quality Issues in Tweet Data`**

* There are rating_denominators greater than 10.
* Numerators with a decimal are parsed incorrectly. (ex: 9.75/10 is parsed as 75/10)

**Define**

Find rows that have a denominator not equal to 10 but also have '/10' somewhere in the text column. Change these rows so that the denominator is 10. Then use find() and extract() to get the numerator (even if it contains a decimal) and update the numerator column with this value. Finally, remove the rest of the columns that have a denominator not equal to 10.

**Code**

In [49]:
# Correctly extract rating numerators even if they have a decimal
tweets_clean['rating_numerator'] = tweets_clean.text.str.extract('(?P<numerator>(?:\d+\.)?\d+)\/(?P<denominator>\d+)', expand=True).numerator.astype(np.float64)

In [50]:
# Replace numerator with correct value if '/10' exists in the text column
tweets_clean['rating_numerator'] = tweets_clean.apply(lambda x: x.text[x.text.find('/10')-1:x.text.find('/10')] if (x.rating_denominator!=10) and ('/10' in x.text) else x.rating_numerator, axis=1).astype(np.float64)
# Replace denominator with 10 if '/10' exists in the text column
tweets_clean['rating_denominator'] = tweets_clean.apply(lambda x: 10 if (x.rating_denominator!=10) and ('/10' in x.text) else x.rating_denominator, axis=1).astype(np.int64)


In [51]:
# drop rows that have a denominator not equal to 10
tweets_clean = tweets_clean[tweets_clean.rating_denominator == 10]

**Test**

In [52]:
# Check example where numerator contains a decimal. The correct rating is 9.75/10 not 75/10.
tweets_clean.loc[695][['text','rating_numerator','rating_denominator']]

text                  This is Logan, the Chow who lived. He solemnly...
rating_numerator                                                   9.75
rating_denominator                                                   10
Name: 695, dtype: object

In [53]:
# Check example of where the rating numerator and denominator was wrong. The correct rating is 9/10 not 1/2.
tweets_clean.loc[2335][['text','rating_numerator','rating_denominator']]

text                  This is an Albanian 3 1/2 legged  Episcopalian...
rating_numerator                                                      9
rating_denominator                                                   10
Name: 2335, dtype: object

In [54]:
# Check that all denominators are equal to 10
tweets_clean[tweets_clean.rating_denominator != 10].shape

(0, 12)

### Storing and Analyzing the Dataset

In [55]:
# To SQLite Database
engine = create_engine('sqlite:///../data/cleaned/we_rate_dogs.db', echo=False)

In [56]:
tweets_clean.to_sql('tweets', con=engine, if_exists='replace', index_label='id')

In [57]:
predictions_clean.to_sql('predictions', con=engine, if_exists='replace', index_label='id')

In [58]:
# To CSV file
tweets_clean.to_csv('../data/cleaned/twitter_archive_master.csv')
predictions_clean.to_csv('../data/cleaned/predictions_master.csv')

In [59]:
# Insight #1: Dog Breeds (with a rank 1 prediction) with the highest average favorite_counts
average_favorite_query = """
select p.prediction as dog_breed, avg(t.favorite_count) as average_favorites
from tweets t
left join predictions p
	on t.tweet_id = p.tweet_id
where p.prediction_rank = 'p1'
	and p.is_dog = 1
group by p.prediction
order by avg(t.favorite_count) desc
"""
average_favorite = pd.read_sql_query(average_favorite_query, engine)
average_favorite.head(10)

dog_breed  average_favorites
0                   saluki       23278.500000
1           french_bulldog       18366.440000
2          giant_schnauzer       16610.000000
3             afghan_hound       16609.666667
4  black-and-tan_coonhound       16399.500000
5    flat-coated_retriever       16194.750000
6      irish_water_spaniel       15751.333333
7          standard_poodle       15227.571429
8         english_springer       15109.444444
9                 cardigan       14720.647059

In [60]:
# Insight #2: Dog Breeds (with a rank 1 prediction) with the highest average retweet_count
average_retweet_query = """
select p.prediction as dog_breed, avg(t.retweet_count) as average_retweets
from tweets t
left join predictions p
	on t.tweet_id = p.tweet_id
where p.prediction_rank = 'p1'
	and p.is_dog = 1
group by p.prediction
order by avg(t.retweet_count) desc
"""
average_retweet = pd.read_sql_query(average_retweet_query, engine)
average_retweet.head(10)

dog_breed  average_retweets
0   standard_poodle       6303.857143
1  english_springer       5672.000000
2      afghan_hound       5655.666667
3        eskimo_dog       5202.333333
4   giant_schnauzer       4860.000000
5            saluki       4859.750000
6    great_pyrenees       4723.307692
7    french_bulldog       4623.000000
8  lakeland_terrier       4592.866667
9           samoyed       4358.675000

In [61]:
# Insight #3: Dog Breeds (with a rank 1 prediction) with the highest average rating numerator
average_rating_query = """
select p.prediction as dog_breed, avg(t.rating_numerator) as rating
from tweets t
left join predictions p
	on t.tweet_id = p.tweet_id
where p.prediction_rank = 'p1'
	and p.is_dog = 1
group by p.prediction
order by avg(t.rating_numerator) desc
"""
average_rating = pd.read_sql_query(average_rating_query, engine)
average_rating.head(10)

dog_breed     rating
0              saluki  12.500000
1              briard  12.333333
2     tibetan_mastiff  12.250000
3      border_terrier  12.142857
4       silky_terrier  12.000000
5  standard_schnauzer  12.000000
6          eskimo_dog  11.777778
7       gordon_setter  11.750000
8        irish_setter  11.750000
9             samoyed  11.700000

### Visualizations in Tableau

<img src="Distribution of Ratings.png">

<img src="10 Most Popular Dog Breed.png">

<img src="Number of Tweets over Time.png">

<img src="Number of Tweets by Dog Type over Time.png">